File: backend/core/responses/userStatsResponse.py
from typing import List

from pydantic import BaseModel

from backend.core.responses.statsHeatmapCellResponse import StatsHeatmapCellResponse
from backend.core.responses.statsMinutesEntryResponse import StatsMinutesEntryResponse
from backend.core.responses.statsRankedItemResponse import StatsRankedItemResponse
from backend.core.responses.statsSummaryResponse import StatsSummaryResponse


class UserStatsResponse(BaseModel):
    summary: StatsSummaryResponse
    minutes: List[StatsMinutesEntryResponse]
    topSongs: List[StatsRankedItemResponse]
    topAlbums: List[StatsRankedItemResponse]
    topArtists: List[StatsRankedItemResponse]
    heatmap: List[StatsHeatmapCellResponse]

File: frontend/packages/shared/src/dto/userStatsResponse.ts
import { z } from "zod";
import { StatsHeatmapCellResponseSchema } from "./statsHeatmapCellResponse";
import { StatsMinutesEntryResponseSchema } from "./statsMinutesEntryResponse";
import { StatsRankedItemResponseSchema } from "./statsRankedItemResponse";
import { StatsSummaryResponseSchema } from "./statsSummaryResponse";

export const UserStatsResponseSchema = z.object({
    summary: z.lazy(() => StatsSummaryResponseSchema),
    minutes: z.array(z.lazy(() => StatsMinutesEntryResponseSchema)),
    topSongs: z.array(z.lazy(() => StatsRankedItemResponseSchema)),
    topAlbums: z.array(z.lazy(() => StatsRankedItemResponseSchema)),
    topArtists: z.array(z.lazy(() => StatsRankedItemResponseSchema)),
    heatmap: z.array(z.lazy(() => StatsHeatmapCellResponseSchema)),
});

export type UserStatsResponse = z.infer<typeof UserStatsResponseSchema>;
